vllm/v1/worker/gpu_model_runner.py

```diff
@@ -1119,6 +1106,21 @@ def execute_model(
             valid_sampled_token_ids = self.rejection_sampler.parse_output(
                 sampled_token_ids, self.input_batch.vocab_size)
 
+        # TODO(woosuk): The following loop can be slow since it iterates over
+        # the requests one by one. Optimize.
+        for i, req_id in enumerate(self.input_batch.req_ids):
```

Comment:
This loop has moved after the cpu sync, it would be better to move it before and collect any request indices to clear in a separate list which can then be processed here.

Reply:
Good point, but I moved here because it needs `valid_sampled_token_ids`, which is generated after the CPU sync. I'll think about whether I can do anything better here, but meanwhile plz let me know if you have any idea in mind.

Reply:
Yes sorry what I meant is move the loop back where it was and just populate a list with the index of any of the requests that should have their `valid_sampled_token_ids cleared`, and then here just have a loop over that list. Most of the time it should be smaller/empty I guess.

Reply:
I tried another approach in https://github.com/vllm-project/vllm/pull/15307/commits/f5ea7246330cf4c7831a1af4b936dc48f99d5fac. Let me know if it looks better to you.